source/common/router/config_impl.h

```diff
@@ -367,8 +367,12 @@ class RouteMatcher {
 
 private:
   const VirtualHostImpl* findVirtualHost(const Http::HeaderMap& headers) const;
+  const VirtualHostImpl* findWildcardVirtualHost(const std::string& host) const;
 
   std::unordered_map<std::string, VirtualHostSharedPtr> virtual_hosts_;
+  // std::greater as a minor optimization to iterate from more to less specific
+  std::map<int64_t, std::unordered_map<std::string, VirtualHostSharedPtr>, std::greater<int64_t>>
```

Comment:
Benchmarking on my workstation, each vector entry costs around 20ns for recall and (string) comparison with a fixed cost of about 25ns. For unordered map, the empty map costs about 65ns and climbs to about 110ns once there are any entries.

All of this boils down to the 0 entry case has a 40ns difference per request, the 1 entry case shows a 65ns difference, and  the break-even point is 4 entries.

The difference is significant in relative terms if we assume the simplest case, but the linear growth is more worrisome to me given that I expect our users may define many wildcard matches and the linear growth seems not awesome. I'd personally rather pay the known fixed cost.

Thoughts?